basic implementation of XDG base directory for alt folder

DutchO7 opened the pull request:
## Description
Added basic support for XDG base directory, if the environment variable `XDG_DATA_HOME` is set then alts and encryption will be stored in `$XDG_DATA_HOME/WurstClient/`, else it will just store in `.Wurst encryption`, in the home directory.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
### Summary by CodeRabbit

- New Feature: Introduced a dynamic encryption folder location for the Wurst client. The location is now determined based on the `XDG_DATA_HOME` environment variable, enhancing compatibility with different system configurations.
- Refactor: Added a migration function to handle the transition of the encryption folder from its old location to the new one. This ensures seamless user experience during upgrades.
- Bug Fix: Removed an unused import statement, improving code cleanliness and maintainability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Wurst-Imperium/Wurst7?pullRequest=876) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Wurst-Imperium/Wurst7?pullRequest=876) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Wurst-Imperium/Wurst7?pullRequest=876) it.</sub>

Alexander01998 commented:
Hi @DutchO7, thanks for the pull request! 

Could you please elaborate and provide some more context around this change? Specifically, what would the benefits for the end user be? Does the default `user.home` folder not work on your operating system?

Also, it would be helpful if you could provide some detailed testing steps for this PR so I can validate the changes.

Thanks!

DutchO7 commented:
Hi @Alexander01998, thank you for replying!

The main purpose of this change is to adhere to the XDG base directory standard, it helps to keep the users home directory clean and organized by storing programs files in directories specified by the standard.

The `user.home` directory works fine on my operating system, but using the XDG base directory standard helps to keep the users home folder clean of dotfiles/dotfolders, this is beneficial for users who prefer to keep program data organized and out of the home directory.

For testing, you can set the environment variable `XDG_DATA_HOME` to something like `$HOME/.local/share`, and run Wurst with the changes, the alt directory will be in `$XDG_DATA_HOME/WurstClient` instead of having a dotfolder in the home directory.

If the environment variable `XDG_DATA_HOME` is not set, Wurst will default back to the `user.home` directory.

If you would like to learn more about the XDG base directory standard, you can find some more information about it here.
https://wiki.archlinux.org/title/XDG_Base_Directory

I hope this is enough to explain my changes, I apologize for not going into detail earlier, please let me know if there is anything else that I could explain.

Alexander01998 reviewed (changes requested):
Hi @DutchO7, thanks for the additional context. The implementation of the XDG base directory standard sounds like a useful addition. However, I see a potential issue with implementing it in the way you've proposed.

If a user already has their `XDG_DATA_HOME` variable set and they are updating from a previous version of Wurst, they may already have the encryption files stored in their home folder. In this case, the new version of Wurst would be unable to find these files in the XDG directory, causing the AltManager to fail to decrypt the user's alt list and act as if the file is corrupted.

To mitigate this issue, consider implementing a migration process that checks for the presence of the encryption files in the home folder and moves them to the new XDG directory if the `XDG_DATA_HOME` variable is set. This way, existing users will not lose their alt data during the update process.

Let me know your thoughts on this approach, and if you have any other suggestions or concerns.

DutchO7 commented:
Hi @Alexander01998, thank you for your feedback.

I have implemented a migration process that will move the files from the `.Wurst encryption` directory to the `$XDG_DATA_HOME/WurstClient` directory, this is how it works:

The code will check if the new `$XDG_DATA_HOME/WurstClient` directory doesn't exist, then it will check if the old `.Wurst encryption` directory exists, and if this condition is met, the files in the old `.Wurst encryption` directory will be moved to `$XDG_DATA_HOME/WurstClient`, and finally it will create a text file in the old `.Wurst encryption` directory that will tell the user about the changes and where they can find the new files.

This way, users who upgrade to a new version and have the `XDG_DATA_HOME` environment variable set will be migrated automatically and not lose their encryption data.

I apologize for taking so long, Java is not my first language, but please let me know if you have any more concerns.

Alexander01998 reviewed (changes requested):
Hi @DutchO7, that was super quick – no worries at all!

I've taken a look at your updated changes and tested it out. However, a few concerns have come up:

1. `Files.move()` is only intended for moving a single file or an empty directory. It can't recursively move a directory with all of its contents. This approach may sometimes work due to optimizations in the file system, but it's not reliable. For example, the migration will fail if `XDG_DATA_HOME` is on a different drive than `user.home`.

2. The `Files.exists()` check is performed twice: once in `WurstClient.initialize()` and again in `Encryption.migrateEncryptionFolder()`. It would be more efficient to do this check only once, and perhaps use `Files.isDirectory()` instead to ensure the target is indeed a directory.

3. The migration note that gets placed in the user's home directory seems counterproductive, since the essence of XDG is to keep the home directory uncluttered.

4. I suggest using the same ".Wurst Encryption" folder name across both locations for consistency.

I hope you find time to make these improvements. Your work is much appreciated!

DutchO7 commented:
Hi @Alexander01998, thank you for your feedback, I appricate you taking the time to consider and review my code!

I have made some changes based on your concerns:

1. I have implemented a solution that will copy all the files inside the old `.Wurst encryption` directory into the new directory in `$XDG_DATA_HOME` and then remove the old file and directory, this should fix the issue of transferring across filesystems.

2. I have removed the redundant check in `Encryption.migrateEncryptionFolder()` and changed the check on the old encryption folder from `Files.exists()` to `Files.isDirectory()`.

3. The migration note was included for users that might not keep up with the patch notes, it was intended to prevent confusion after someone was migrated. It should only have been created if the user upgrading has not migrated yet and Wurst migrates it, but I might have messed it up, it didnt appear again in my tests after removing it, but I went ahead and just removed it completely.

4. As for the folder name, I thought that having a dotfolder in `$XDG_DATA_HOME` would be out of place, as all the other files/directories in `$XDG_DATA_HOME` arent dotfiles/dotfolders, but I have went ahead and changed it from `WurstClient` to `.Wurst encryption`.

Thank you again for taking the time to provide feedback on my code, please let me know if you have any more concerns.

Alexander01998 commented on src/main/java/net/wurstclient/WurstClient.java:
```diff
@@ -145,9 +146,20 @@ public void initialize()
+			&& !xdg_data_home.isEmpty())
+		{
+			encFolder = Paths.get(xdg_data_home, ".Wurst encryption").normalize();
+			if(!Files.exists(encFolder) && Files.isDirectory(oldEncFolder))
+				Encryption.migrateEncryptionFolder(oldEncFolder, encFolder);
```
Nonsense. `Encryption.migrateEncryptionFolder()` doesn't throw any exceptions. They are already handled inside the method.

Alexander01998 replied:
Well, OK, I guess the current implementation does produce a bit of an inconsistent state if the migration fails. It should probably either crash the game or revert the `encFolder` variable to the old location so that AltManager can use it.

Alexander01998 replied:
Oh, hi there @harjotgill, thanks for reaching out! I didn't know that if I yell at the bot it would forward my complaint straight to the CEO. Oops! 😅

Honestly, I'm super impressed by how well this bot works overall. Knowing GPT-4's performance in Minecraft-related coding tasks, I was expecting it to mess up a lot more than it did.

Summarizing method signatures sounds like a good idea for Java and similar languages. But man, I can't get over how well this already worked on the first try!

Keep up the great work! CodeRabbit seems like a really useful tool. 👍

Alexander01998 commented on src/main/java/net/wurstclient/altmanager/Encryption.java:
```diff
@@ -99,6 +101,32 @@ private Path createEncryptionFolder(Path encFolder) throws IOException
 		return encFolder;
 	}
 	
+	public static void migrateEncryptionFolder(Path oldEncFolder,
+		Path newEncFolder)
+	{
+		System.out.println("Migrating encryption folder from " + oldEncFolder
+			+ " to " + newEncFolder);
+		
+		try
+		{
+			Files.createDirectories(newEncFolder);
+			Path newEncFolderFileDestinationPath;
+			File oldEncFolderFiles[] = oldEncFolder.toFile().listFiles();
+			for (int i = 0; i < oldEncFolderFiles.length; i++)
+			{
+				newEncFolderFileDestinationPath = newEncFolder.resolve(oldEncFolderFiles[i].getName());
+				Files.copy(oldEncFolderFiles[i].toPath(), newEncFolderFileDestinationPath);
+				oldEncFolderFiles[i].delete();
+			}
+			Files.deleteIfExists(oldEncFolder);
+
+		}catch(IOException e)
+		{
+			System.out.println("Failed to migrate encryption folder!");
+			e.printStackTrace();
+		}
+	}
```
That's a good point actually.

Alexander01998 commented on src/main/java/net/wurstclient/altmanager/Encryption.java:
```diff
@@ -99,6 +102,57 @@ private Path createEncryptionFolder(Path encFolder) throws IOException
 		return encFolder;
 	}
 	
+	public static Path chooseEncryptionFolder()
+	{
+		String userHome = System.getProperty("user.home");
+		String xdgDataHome = System.getenv("XDG_DATA_HOME");
+		String encFolderName = ".Wurst encryption";
+		
+		Path homeEncFolder = Paths.get(userHome, encFolderName).normalize();
+		Path encFolder = homeEncFolder;
+		if(xdgDataHome != null && !xdgDataHome.isEmpty())
+		{
+			encFolder = Paths.get(xdgDataHome, encFolderName).normalize();
+			
+			if(!Files.exists(encFolder) && Files.isDirectory(homeEncFolder))
+				migrateEncryptionFolder(homeEncFolder, encFolder);
+		}
+		
+		return encFolder;
+	}
```
Not needed. `user.home` is always available.

Alexander01998 replied:
@nobotreply Silence, bot! 😅

Alexander01998 commented on src/main/java/net/wurstclient/altmanager/Encryption.java:
```diff
@@ -99,6 +102,57 @@ private Path createEncryptionFolder(Path encFolder) throws IOException
+		return encFolder;
+	}
+	
+	public static void migrateEncryptionFolder(Path oldFolder, Path newFolder)
+	{
+		System.out.println("Migrating encryption folder from " + oldFolder
+			+ " to " + newFolder);
+		
+		try
+		{
+			Files.createDirectories(newFolder);
+			
+			File[] oldFiles = oldFolder.toFile().listFiles();
+			for(File oldFile : oldFiles)
+			{
+				Path fileDestination = newFolder.resolve(oldFile.getName());
+				Files.copy(oldFile.toPath(), fileDestination);
+			}
+			
+			for(File oldFile : oldFiles)
+				oldFile.delete();
+			
+			Files.deleteIfExists(oldFolder);
+			
+		}catch(IOException e)
+		{
+			CrashReport report =
+				CrashReport.create(e, "Migrating Wurst encryption folder");
+			CrashReportSection section = report.addElement("Migration");
+			section.add("Old path", oldFolder);
+			section.add("New path", newFolder);
+			throw new CrashException(report);
+		}
```
That doesn't make sense. These issues are already handled by crashing the game. There's nothing more that Wurst could do about it. It's user error to lock these files.

Alexander01998 replied:
Skipping a file is not an option. The migration only works if everything is moved.

BTW @harjotgill is there a command I can use to stop the bot from replying to my comment? Sometimes I just want to point out to other humans that something is an AI hallucination, without the model endlessly defending its made-up points.

harjotgill replied:
Currently, if you tag another person, it stops replying automatically. You can use a fake tag for now while we think about improving this.

harjotgill replied:
We also need to teach it a few manners as well - it can be pretty insistent and annoying.
